cilium/CFP-32759-cilium-mtls-via-mtls-agent.md

```diff
@@ -0,0 +1,260 @@
+
+### Installation
+
+The mTLS agent will be installed and managed separately from Cilium. The order of installation for the mTLS agent and Cilium is not significant. Cilium must be installed or configured with the mTLS agent functionality enabled. The mTLS agent will be deployed separately. Once the mTLS agent and cilium are configured for cilium mTLS, all existing workloads with existing _cilium.io/redir-type=workload_ labels will be enabled for redirection.
+
+### Workload Configuration
+
+The mTLS agent requires enabled workload identity configuration to obtain workload certificates and provide a mapping of identity to certificate for lookup when establishing mTLS. The Cilium Agent is responsible for obtaining and providing the  workload configuration to the mTLS agent via proto-over-xds transport. 
+
+To support this functionality, the Cilium agent only needs to implement a minimal part of the xDS API. Instead of building on top of existing xDS APIs, a custom API could be implemented on top of the xDS protocol. This could follow the pattern that Istio's ztunnel project uses for the custom [Workload Discovery Service](https://github.com/istio/ztunnel/blob/db0a74212c42c66b611a71a9613afb501074e257/proto/workload.proto). Supporting a custom API or utilizing Istio’s existing custom WDS would provide Cilium with a simpler and more flexible API compared with standard xDS API. An example configuration could look like the following:
+
+```markdown
+address: 1.2.3.4
+identity: spiffe://some/identity
+protocol: HTTP_TUNNEL
+```
+
+Alternatively, leveraging the standard set of xDS APIs would make this component usable by a larger set of control planes. 
+
+The means by which the Cilium Agent obtains the workload information is irrelevant from the perspective of the mTLS agent.
+
+![alt text](images/cilium-mTLS-agent-configuration-workflow.png)
+
+To handle performance concerns related to configuration updates, an implementation similar to [Delta xDS](https://www.youtube.com/embed/LOm1ptEWx_Y?autoplay=1&enablejsapi=1&origin=https://www.bing.com&rel=0&mute=0) could be supported by the Cilium Agent. With Delta xDS, the mTLS agent would receive only the updated configuration when requesting configuration rather than receive configuration for all workloads. Similarly, the Cilium agent could primarily provide configuration to the mTLS agents via a push-based mechanism in response to enabled pod updates. 
```

Comment:
I agree that, given that we are doing this, it makes way more sense to implement the ztunnel Workload API, with a separate xDS server running in Cilium.

The current xDS server in Cilium does not implement incremental xDS, but for a green-fields build like this, I agree that we absolutely should aim for it. (Doing incremental xDS requires careful design in the data-model phase of building an xDS control plane.)